monai/apps/pathology/transforms/stain/array.py

```diff
@@ -0,0 +1,194 @@
+# Copyright 2020 - 2021 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+import numpy as np
+
+from monai.transforms.transform import Transform
+
+
+class ExtractHEStains(Transform):
+    """Class to extract a target stain from an image, using stain deconvolution (see Note).
+
+    Args:
+        tli: transmitted light intensity. Defaults to 240.
+        alpha: tolerance in percentile for the pseudo-min (alpha percentile)
+            and pseudo-max (100 - alpha percentile). Defaults to 1.
+        beta: absorbance threshold for transparent pixels. Defaults to 0.15
+        max_cref: reference maximum stain concentrations for Hematoxylin & Eosin (H&E).
+            Defaults to None.
+
+    Note:
+        For more information refer to:
+        - the original paper: Macenko et al., 2009 http://wwwx.cs.unc.edu/~mn/sites/default/files/macenko2009.pdf
+        - the previous implementations:
+
+          - MATLAB: https://github.com/mitkovetta/staining-normalization
+          - Python: https://github.com/schaugf/HEnorm_python
+    """
+
+    def __init__(
+        self,
+        tli: float = 240,
+        alpha: float = 1,
+        beta: float = 0.15,
+        max_cref: Optional[np.ndarray] = None,
+    ) -> None:
+        self.tli = tli
+        self.alpha = alpha
+        self.beta = beta
+
+        self.max_cref = max_cref
+        if self.max_cref is None:
+            self.max_cref = np.array([1.9705, 1.0308])
+
+    def _deconvolution_extract_stain(self, image: np.ndarray) -> np.ndarray:
+        """Perform Stain Deconvolution and return stain matrix for the image.
+
+        Args:
+            img: uint8 RGB image to perform stain deconvolution of
+
+        Return:
+            he: H&E absorbance matrix for the image (first column is H, second column is E, rows are RGB values)
+        """
+        # reshape image and calculate absorbance
+        image = image.reshape((-1, 3))
+        image = image.astype(np.float32) + 1.0
+        absorbance = -np.log(image.clip(max=self.tli) / self.tli)
```

Comment:
do we need some validation of the input of np.log?

Reply:
That's a good point! We should check if the input is intensity (0-255) and not any image or other scales. Thanks @wyli!

Reply:
Now it checks the image values to be between 0 and 255.